File: 백준/Gold/9663. N－Queen/N－Queen.java
import java.util.*;

import static java.lang.Math.abs;

public class Main {

    static int N;
    static int count=0;
    static boolean[] column, diag1, diag2;
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        N= sc.nextInt();
        column = new boolean[N];
        diag1 = new boolean[2*N-1];
        diag2 = new boolean[2*N-1];

        dfs(0);

        System.out.println(count);
    }

    public static void dfs(int row){
        if(row==N){
            count++;
            return;
        }

        for(int col=0;col<N;col++){
            if(!column[col] && !diag1[row+col] && !diag2[row-col+(N-1)]){
                column[col] = diag1[row+col] = diag2[row-col+(N-1)] = true;

                dfs(row+1);

                column[col] = diag1[row+col] = diag2[row-col+(N-1)] = false;
            }
        }
    }
}

File: 백준/Gold/9663. N－Queen/N－Queen.py
def solve(row, col, diag1, diag2, N):
    if row == N:
        return 1
    count = 0
    available_positions = ((1 << N) - 1) & ~(col | diag1 | diag2)
    while available_positions:
        position = available_positions & -available_positions  # 최하위 비트 선택
        available_positions -= position  # 해당 위치 사용
        count += solve(row + 1, col | position, (diag1 | position) << 1, (diag2 | position) >> 1, N)
    return count

N = int(input())
print(solve(0, 0, 0, 0, N))